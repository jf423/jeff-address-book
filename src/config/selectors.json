{
    "INDEX_PAGE": {
        "INDEX_CONTAINER": {
            "ID": "INDEX_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='INDEX_CONTAINER']",
            "EX": "[data-selector-id='INDEX_CONTAINER']",
            "PARAM": ""
        },
        "SEARCH_CONTAINER": {
            "ID": "SEARCH_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_CONTAINER']",
            "EX": "[data-selector-id='SEARCH_CONTAINER']",
            "PARAM": ""
        },
        "SEARCH_INPUT": {
            "ID": "SEARCH_INPUT",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_INPUT']",
            "EX": "[data-selector-id='SEARCH_INPUT']",
            "PARAM": ""
        },
        "SEARCH_CLEAR": {
            "ID": "SEARCH_CLEAR",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_CLEAR']",
            "EX": "[data-selector-id='SEARCH_CLEAR']",
            "PARAM": ""
        },
        "SEARCH_BUTTON": {
            "ID": "SEARCH_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='SEARCH_BUTTON']",
            "EX": "[data-selector-id='SEARCH_BUTTON']",
            "PARAM": ""
        },
        "USER_CONTAINER": {
            "ID": "USER_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CONTAINER']",
            "EX": "[data-selector-id='USER_CONTAINER']",
            "PARAM": ""
        },
        "USER_CONTENT": {
            "ID": "USER_CONTENT",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CONTENT']",
            "EX": "[data-selector-id='USER_CONTENT']",
            "PARAM": ""
        },
        "USER_CATALOG_END": {
            "ID": "USER_CATALOG_END",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CATALOG_END']",
            "EX": "[data-selector-id='USER_CATALOG_END']",
            "PARAM": ""
        },
        "USER_CELL_CONTAINER": {
            "ID": "USER_CELL_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1']",
            "PARAM": ""
        },
        "USER_CELL_CONTENT": {
            "ID": "USER_CELL_CONTENT",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_CONTENT']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_CONTENT']",
            "PARAM": ""
        },
        "USER_CELL_AVATAR": {
            "ID": "USER_CELL_AVATAR",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_AVATAR']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_AVATAR']",
            "PARAM": ""
        },
        "USER_CELL_NAME": {
            "ID": "USER_CELL_NAME",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_NAME']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_NAME']",
            "PARAM": ""
        },
        "USER_CELL_NATIONALITY": {
            "ID": "USER_CELL_NATIONALITY",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_NATIONALITY']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_NATIONALITY']",
            "PARAM": ""
        },
        "USER_CELL_USERNAME": {
            "ID": "USER_CELL_USERNAME",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_USERNAME']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_USERNAME']",
            "PARAM": ""
        },
        "USER_CELL_EMAIL": {
            "ID": "USER_CELL_EMAIL",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_CELL_CONTAINER_{INDEX}'] [data-selector-id='USER_CELL_EMAIL']",
            "EX": "[data-selector-id='USER_CELL_CONTAINER_1'] [data-selector-id='USER_CELL_EMAIL']",
            "PARAM": ""
        },
        "USER_MODAL_CONTAINER": {
            "ID": "USER_MODAL_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CONTAINER']",
            "EX": "[data-selector-id='USER_MODAL_CONTAINER']",
            "PARAM": ""
        },
        "USER_MODAL_CLOSE_BUTTON": {
            "ID": "USER_MODAL_CLOSE_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CLOSE_BUTTON']",
            "EX": "[data-selector-id='USER_MODAL_CLOSE_BUTTON']",
            "PARAM": ""
        },
        "USER_MODAL_AVATAR": {
            "ID": "USER_MODAL_AVATAR",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_AVATAR']",
            "EX": "[data-selector-id='USER_MODAL_AVATAR']",
            "PARAM": ""
        },
        "USER_MODAL_NAME": {
            "ID": "USER_MODAL_NAME",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_NAME']",
            "EX": "[data-selector-id='USER_MODAL_NAME']",
            "PARAM": ""
        },
        "USER_MODAL_STREET": {
            "ID": "USER_MODAL_STREET",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_STREET']",
            "EX": "[data-selector-id='USER_MODAL_STREET']",
            "PARAM": ""
        },
        "USER_MODAL_CITY": {
            "ID": "USER_MODAL_CITY",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CITY']",
            "EX": "[data-selector-id='USER_MODAL_CITY']",
            "PARAM": ""
        },
        "USER_MODAL_STATE": {
            "ID": "USER_MODAL_STATE",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_STATE']",
            "EX": "[data-selector-id='USER_MODAL_STATE']",
            "PARAM": ""
        },
        "USER_MODAL_POSTCODE": {
            "ID": "USER_MODAL_POSTCODE",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_POSTCODE']",
            "EX": "[data-selector-id='USER_MODAL_POSTCODE']",
            "PARAM": ""
        },
        "USER_MODAL_PHONE": {
            "ID": "USER_MODAL_PHONE",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_PHONE']",
            "EX": "[data-selector-id='USER_MODAL_PHONE']",
            "PARAM": ""
        },
        "USER_MODAL_CELL": {
            "ID": "USER_MODAL_CELL",
            "TYPE": "css",
            "VAL": "[data-selector-id='USER_MODAL_CELL']",
            "EX": "[data-selector-id='USER_MODAL_CELL']",
            "PARAM": ""
        }
    },
    "SETTINGS_PAGE": {
        "SETTINGS_CONTAINER": {
            "ID": "SETTINGS_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='SETTINGS_CONTAINER']",
            "EX": "[data-selector-id='SETTINGS_CONTAINER']",
            "PARAM": ""
        },
        "NATIONALITIES_CONTAINER": {
            "ID": "NATIONALITIES_CONTAINER",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CONTAINER']",
            "EX": "[data-selector-id='NATIONALITIES_CONTAINER']",
            "PARAM": ""
        },
        "NATIONALITIES_BACK_BUTTON": {
            "ID": "NATIONALITIES_BACK_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_BACK_BUTTON']",
            "EX": "[data-selector-id='NATIONALITIES_BACK_BUTTON']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_CH": {
            "ID": "NATIONALITIES_CHECKBOX_CH",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_CH']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_CH']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_ES": {
            "ID": "NATIONALITIES_CHECKBOX_ES",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_ES']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_ES']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_FR": {
            "ID": "NATIONALITIES_CHECKBOX_FR",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_FR']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_FR']",
            "PARAM": ""
        },
        "NATIONALITIES_CHECKBOX_GB": {
            "ID": "NATIONALITIES_CHECKBOX_GB",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_CHECKBOX_GB']",
            "EX": "[data-selector-id='NATIONALITIES_CHECKBOX_GB']",
            "PARAM": ""
        },
        "NATIONALITIES_SAVE_BUTTON": {
            "ID": "NATIONALITIES_SAVE_BUTTON",
            "TYPE": "css",
            "VAL": "[data-selector-id='NATIONALITIES_SAVE_BUTTON']",
            "EX": "[data-selector-id='NATIONALITIES_SAVE_BUTTON']",
            "PARAM": ""
        }
    }
}