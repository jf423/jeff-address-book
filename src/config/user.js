export const USERS_CONFIG = {
    NUM_COLUMNS: 4,
    BATCH_NUM: 50,
    MAXIMUM_COUNT: 150,
    ITEM_WIDTH: 250,
    ITEM_HEIGHT: 300
};

export const END_POINT = 'https://randomuser.me/api/';

export const DEFAULT_NATIONALITIES = ['ch', 'es', 'fr', 'gb'];
