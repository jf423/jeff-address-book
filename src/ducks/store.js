import { useMemo } from 'react';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

import reducer, { initialState } from './modules/users';

let store;

export function initStore(preloadedState = initialState) {
    return createStore(
        reducer,
        preloadedState,
        composeWithDevTools(applyMiddleware())
    );
}

export function initializeStore(preloadedState) {
    const _store = store || initStore(preloadedState);
    if (typeof window === 'undefined') return _store;
    if (!store) store = _store;

    return _store;
}

export function useStore(state) {
    return useMemo(() => initializeStore(state), [state]);
}
