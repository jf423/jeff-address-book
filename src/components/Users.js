import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { FixedSizeGrid as Grid } from 'react-window';
import InfiniteLoader from 'react-window-infinite-loader';
import AutoSizer from 'react-virtualized-auto-sizer';
import styled from 'styled-components';

import { IconButton, Snackbar } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';

import { flexColStartCenter } from '@/style/flex.css.js';

import { updateUserDataAction } from '@/ducks/modules/users.js';
import { USERS_CONFIG, END_POINT } from '@/config/user.js';
import { INDEX_PAGE } from '@/config/selectors.json';

import useColumnCount from '@/hooks/useColumnCount.js';

import { getFilterUser } from '@/utils/users.js';

import Search from './Search‌.js';
import UserModal from './UserModal.js';
import UserCell from './UserCell.js';

const {
    NUM_COLUMNS,
    BATCH_NUM,
    MAXIMUM_COUNT,
    ITEM_WIDTH,
    ITEM_HEIGHT
} = USERS_CONFIG;
const {
    USER_CONTAINER,
    USER_CONTENT,
    USER_CATALOG_END
} = INDEX_PAGE;

const LOADING_COLUMNS = new Array(NUM_COLUMNS).fill({});

const Users = () => {
    const dispatch = useDispatch();
    const { searchText, nationalities } = useSelector(state => state);
    const userData = useSelector(state => getFilterUser(state.userData, nationalities, searchText));
    const [selectedIndex, setSelectedIndex] = useState(null);
    const [isScrollToEnd, setScrollToEnd] = useState(false);
    const [isDisableSnackbar, setDisableSnackbar] = useState(false);
    const userCount = userData.length;
    const items = (userCount < MAXIMUM_COUNT && !searchText) ? userData.concat(LOADING_COLUMNS) : userData;

    const onCloseSnackbar = () => {
        setDisableSnackbar(true);
    };
    const onLoadMore = () => {
        if (searchText) return;

        return fetch(`${END_POINT}?results=${BATCH_NUM}&nat=${nationalities.join()}`)
            .then(res => res.json())
            .then(({ results }) => {
                dispatch(updateUserDataAction(userData.concat(results)));
            });
    };
    const isItemLoaded = index => {
        if (userCount === MAXIMUM_COUNT && userCount === index + 1) {
            setScrollToEnd(true);
        }
        return index < userCount;
    };

    return (
        <Container data-selector-id={USER_CONTAINER.ID}>
            <SearchHeader />
            <UsersContent data-selector-id={USER_CONTENT.ID}>
                <AutoSizer>
                    {({ width, height }) => (
                        <InfiniteLoader
                            itemCount={items.length}
                            isItemLoaded={isItemLoaded}
                            loadMoreItems={onLoadMore}
                        >
                            {({ onItemsRendered, ref }) => (
                                <Grid
                                    ref={ref}
                                    width={width}
                                    height={height}
                                    columnWidth={ITEM_WIDTH}
                                    rowHeight={ITEM_HEIGHT}
                                    columnCount={NUM_COLUMNS}
                                    rowCount={Math.ceil(items.length / NUM_COLUMNS)}
                                    itemData={{ items, setSelectedIndex }}
                                    onItemsRendered={gridProps => onItemsRendered({
                                        overscanStartIndex: gridProps.overscanRowStartIndex * NUM_COLUMNS,
                                        overscanStopIndex: gridProps.overscanRowStopIndex * NUM_COLUMNS,
                                        visibleStartIndex: gridProps.visibleRowStartIndex * NUM_COLUMNS,
                                        visibleStopIndex: gridProps.visibleRowStopIndex * NUM_COLUMNS
                                    })}
                                >
                                    {UserCell}
                                </Grid>
                            )}
                        </InfiniteLoader>
                    )}
                </AutoSizer>
            </UsersContent>
            <UserModal open={Boolean(userData[selectedIndex])} selectedUser={userData[selectedIndex]} setSelectedIndex={setSelectedIndex} />
            <Snackbar
                data-selector-id={USER_CATALOG_END.ID}
                anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
                open={!isDisableSnackbar && isScrollToEnd}
                message="End‌ ‌of‌ ‌users‌ ‌catalog!"
                action={
                    <IconButton
                        size="small"
                        aria-label="close"
                        color="inherit"
                        onClick={onCloseSnackbar}
                    >
                        <CloseIcon fontSize="small" />
                    </IconButton>
                }
            />
        </Container>
    );
};

export default Users;

const Container = styled.div`
    width: ${ITEM_WIDTH * NUM_COLUMNS}px;
    height: 100%;
    ${flexColStartCenter};
`;

const SearchHeader = styled(Search)`
    width: 100%;
    height: 40px;
    margin-top: 15px;
`;

const UsersContent = styled.div`
    width: 100%;
    height: calc(100% - 75px);
`;
