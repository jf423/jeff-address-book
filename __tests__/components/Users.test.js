import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { toBeInTheDocument, toHaveTextContent, toHaveValue } from '@testing-library/jest-dom/matchers';

import { initStore } from '@/ducks/store.js';

import { INDEX_PAGE } from '@/config/selectors.json';
import { DEFAULT_NATIONALITIES } from '@/config/user.js';

import Users from '@/components/Users.js';

import DATA from '../mockUsersData.json';

expect.extend({ toBeInTheDocument, toHaveTextContent, toHaveValue });

const initialState = {
    userData: DATA.results,
    nationalities: DEFAULT_NATIONALITIES,
    searchText: ''
};
const mockDispatch = jest.fn();
const {
    USER_CONTAINER,
    USER_CONTENT
} = INDEX_PAGE;

describe('User component test', () => {
    let user;
    const store = initStore(initialState);
    store.dispatch = mockDispatch;
    const renderUser = props => (
        <Provider store={store}>
            <Users {...props} />
        </Provider>
    );

    beforeEach(() => {
        user = render(renderUser());
    });

    afterEach(() => {
        user.unmount();
    });

    test('should display as expected', () => {
        const userContainer = user.getByTestId(USER_CONTAINER.ID);
        const userContent = user.getByTestId(USER_CONTENT.ID);

        expect(userContainer).toBeInTheDocument();
        expect(userContent).toBeInTheDocument();
        expect(user).toMatchSnapshot();
    });
});
