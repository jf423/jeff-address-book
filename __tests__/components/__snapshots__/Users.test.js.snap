// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component test should display as expected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Users__Container-sc-15n28nk-0 fScasL"
        data-selector-id="USER_CONTAINER"
      >
        <div
          class="Search__Container-sc-1j9mi1x-0 cWJnpy Users__SearchHeader-sc-15n28nk-1 gQMvCT"
          data-selector-id="SEARCH_CONTAINER"
        >
          <div
            class="MuiInputBase-root Search__SearchInput-sc-1j9mi1x-1 iNXuXD"
            data-selector-id="SEARCH_INPUT"
            title="Searching will pause‌ ‌the‌ ‌loading‌ ‌mechanism‌ due to the limitation of Random User API."
          >
            <input
              aria-label="search user"
              class="MuiInputBase-input"
              placeholder="Search User"
              type="text"
              value=""
            />
          </div>
          <button
            aria-label="clear"
            class="MuiButtonBase-root MuiIconButton-root Search__CloseButton-sc-1j9mi1x-2 hSdNud"
            data-selector-id="SEARCH_CLEAR"
            tabindex="0"
            type="button"
            visible="hidden"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="search"
            class="MuiButtonBase-root MuiIconButton-root"
            data-selector-id="SEARCH_BUTTON"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <hr
            class="MuiDivider-root MuiDivider-vertical"
          />
          <button
            aria-label="settings"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="Users__UsersContent-sc-15n28nk-2 eGaVlM"
          data-selector-id="USER_CONTENT"
        >
          <div
            style="overflow: visible; height: 0px; width: 0px;"
          />
          <div
            class="resize-triggers"
          >
            <div
              class="expand-trigger"
            >
              <div
                style="width: 1px; height: 1px;"
              />
            </div>
            <div
              class="contract-trigger"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Users__Container-sc-15n28nk-0 fScasL"
      data-selector-id="USER_CONTAINER"
    >
      <div
        class="Search__Container-sc-1j9mi1x-0 cWJnpy Users__SearchHeader-sc-15n28nk-1 gQMvCT"
        data-selector-id="SEARCH_CONTAINER"
      >
        <div
          class="MuiInputBase-root Search__SearchInput-sc-1j9mi1x-1 iNXuXD"
          data-selector-id="SEARCH_INPUT"
          title="Searching will pause‌ ‌the‌ ‌loading‌ ‌mechanism‌ due to the limitation of Random User API."
        >
          <input
            aria-label="search user"
            class="MuiInputBase-input"
            placeholder="Search User"
            type="text"
            value=""
          />
        </div>
        <button
          aria-label="clear"
          class="MuiButtonBase-root MuiIconButton-root Search__CloseButton-sc-1j9mi1x-2 hSdNud"
          data-selector-id="SEARCH_CLEAR"
          tabindex="0"
          type="button"
          visible="hidden"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-label="search"
          class="MuiButtonBase-root MuiIconButton-root"
          data-selector-id="SEARCH_BUTTON"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <hr
          class="MuiDivider-root MuiDivider-vertical"
        />
        <button
          aria-label="settings"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="Users__UsersContent-sc-15n28nk-2 eGaVlM"
        data-selector-id="USER_CONTENT"
      >
        <div
          style="overflow: visible; height: 0px; width: 0px;"
        />
        <div
          class="resize-triggers"
        >
          <div
            class="expand-trigger"
          >
            <div
              style="width: 1px; height: 1px;"
            />
          </div>
          <div
            class="contract-trigger"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
